README

			 Nagios SNMP Plugins
			 ===================

What is it?
-----------

These are two plugins to be used with Nagios. They should still work
with NetSaint, but as this is now really ancient, I don't support
Netsaint any longer. You also need the Net SNMP Daemon in a reasonable
new version.

Nagios:   	http://www.nagios.org/
Net-SNMPD: 	http://www.net-snmp.org/


Who wrote it?
-------------

These programs are (C) 2002-2008 by
Henning P. Schmiedehausen

(C) 2010-2013 Elan Ruusamäe <glen@delfi.ee>

Distributed under GNU GENERAL PUBLIC LICENSE v2 
(See file COPYING in the distribution).


Where to get it?
----------------

Homepage: https://github.com/glensc/nagios-snmp-plugins
Release downloads: https://github.com/glensc/nagios-snmp-plugins/releases

Acknowlegdements & Thanks
-------------------------

Elan Ruusamäe <glen at pld-linux.org>
   - enhanced reporting for the check_snmp_proc plugin. Also
     the only patch I ever got from a distribution vendor. :-)

Jose Pedro Oliveira <jpo at di.uminho.pt> 

   - for maintaining the Fedora Extras RPM; I rolled a number of
   things from that spec file back into the "official" one.

Oliver Faenger <oliver.faenger at cityweb.de>
   - representing many many people who reported and sent me patches
     for the gcc4 compilation issue. 

Mark Janssen <maniac at maniac.nl>
   - patch that displays the checked devices and processes

Henning P. Schmiedehausen
   - original code author


All the Linux distributions who included that little hack into their
package lists. BTW: I'd appreciate if you drop me a quick line when
you include nagios-snmp-plugins into a distribution. So far I know of:

- Fedora 7
- Fedora Extras 
- PLD Linux


What can these plugins do?
--------------------------

NET-SNMP gives you an unique feature: It makes checks for running
processes and/or disk size configurable in the SNMP daemon and
queryable via SNMP. So you can do

proc  syslogd
proc  klogd

disk  /    5%
disk  /usr 10%


in your snmpd.local.conf and have the snmp daemon check that you
always have free space on your root and usr file system and that you
have syslogd and klogd processes running.

Querying this from a monitoring tool however, is very cumbersome. You
need to check what disks are there, then query error status.

These two plugins does all the work for you.


Where do the plugins run?
-------------------------

Well, on Linux. :-) That's where I wrote them. They should compile on
any other *x like Operating system, too. If you get them to compile,
please let me know.

I compiled and tested these plugins on

RedHat Linux 9
Fedora Core 1
Fedora Core 3
Fedora Core 4
Fedora Core 5
Fedora Core 6
RedHat Enterprise Linux 3
RedHat Enterprise Linux 4
RedHat Enterprise Linux 5
CentOS 5

using various nagios versions between 1.2 and 2.9.

Users reported success on various SuSE Linux and SLES
versions. Compilation on *BSD seems to be possible if you have a GNU
getline library.


How do I build and install?
---------------------------

- Unpack the distribution.

- Run

    aclocal
    autoheader
    automake --add-missing
    autoconf
    ./configure
    make

  or simply

    ./build.sh

- you should now have two binaries:  check_snmp_disk and check_snmp_proc

- copy them into your Nagios plugins directory.


How do I configure the plugins?
-------------------------------

Using Nagios there are two ways to configure the plugins:

The old way:

- add the following two lines to your /etc/nagios/command.cfg:

command[check_snmp_disk]=/usr/lib/nagios/plugins/check_snmp_disk -H $HOSTADDRESS$ -C $ARG1$
command[check_snmp_proc]=/usr/lib/nagios/plugins/check_snmp_proc -H $HOSTADDRESS$ -C $ARG1$

  Replace /usr/lib/nagios/plugins/ with the directory where you copied
  the binaries!

The new way:

- add the following lines to your /etc/nagios/checkcommands.cfg

# 'check_snmp_disk' command definition
define command	{
	command_name	check_snmp_disk
	command_line	$USER1$/check_snmp_disk -H $ARG1$ -C $USER3$
}


# 'check_snmp_proc' command definition
define command	{
	command_name	check_snmp_proc
	command_line	$USER1$/check_snmp_proc -H $ARG1$ -C $USER3$
}

- The Community password should be defined in your resources.cfg If you don't use
  $USER3$, adjust the definition above.


How do I use the plugins?
-------------------------

First, select something to check. You find a detailed description in
man snmpd.conf(5).

I, e.g. use (in /etc/snmpd/snmpd.local.conf under RedHat/Fedora):

#
# SNMP checked disks
#
disk	/		5%
disk	/usr		5%
disk	/var		10%
disk	/boot		1%
disk	/mnt/disk0	10%
#
# SNMP checked Processes
#
proc	crond
proc	ntpd
proc	sshd
proc	syslogd
proc	klogd

Which assures that all my disks do not fill up and that I have the
most important processes running on my boxes.

You need a configured snmpd with at least a read community configured.

Now try your configuration: Run the disk plugin manually:

check_snmp_disk -H <host to check> -C <community>

It should report either

"Checked <xxx> disks."  and a return code of 0 if everything is fine
(<xxx> should be the number of disks configured in the snmp.local.conf
file),

or

"<diskname>: less than <xxx>% free (= <yyy>%) (<devicename>)" and a
return code of 2 if there is a problem (diskname and devicename depend
on the disk, <xxx> is the threshold configured and <yyy> is the actual
free space).


The same testing can be done with the check_snmp_proc plugin.

Now configure your monitor program to monitor a host (in hosts.cfg):

The old way:

service[myhost]=DISK;0;24x7;3;5;1;admins;120;24x7;1;1;0;;check_snmp_disk!<community>
service[myhost]=PROC;0;24x7;3;5;1;admins;120;24x7;1;1;0;;check_snmp_proc!<community>

where <community is your read-community from above) and you're
set. The ";0;24x7;3;5;1;admins;120;24x7;1;1;0;;" parameters should be
of course your very own configuration. :-)

The new way:

# Service definition
define service	{
	use				generic-service

	host_name			myhost
	service_description		Disk Space
	check_command			check_snmp_disk!myhost
}

# Service definition
define service	{
	use				generic-service

	host_name			myhost
	service_description		Running processes
	check_command			check_snmp_proc!myhost
}

Of course you should replace "myhost" with your host. 


What is left to do?
-------------------

- Hard/Soft Thresholds 

- If multiple failures occur, only the first is reported
  (Nagios restriction) and every further error only when the
  first is fixed.


Release History
---------------

see NEWS FILE


NEWS
Version 1.3.1 - Released 2013-10-01
* Avoid using newlines in check_snmp_disk when reporting errors

Version 1.3 - Released 2010-10-30
* New maintainer: Elan Ruusamäe
* Added check_snmp_swap
* Added check_snmp_load
* Add -p in check_snmp_disk to output mountpoints instead of devices in -l mode

Version 1.2 - Released 2008-01-27
* Added patch to display disks and processes, 
  contributed by Mark Janssen

Version 1.1 - Released 2007-08-27
* Fixed annoying compile problem with distributions post-Fedora Core 1
* Add check_snmp_proc formatting contributed by Elan Ruusamäe

Version 1.0 - Released 2004-03-29
* Renamed to nagios-snmp-plugins
* Added autoconf support

Version 0.1 - Released 2002-01-27


snmp_common.c
/*************************************************************************
 *
 * plugins for Nagios
 *
 * (C) 2002-2008
 *   Henning P. Schmiedehausen
 *   INTERMETA - Gesellschaft fuer Mehrwertdienste mbH
 *   Hutweide 15
 *   D-91054 Buckenhof
 * (C) 2010-2013
 *   Elan Ruusamäe, <glen@delfi.ee>
 *
 *************************************************************************
 *
 * Common code
 *
 *************************************************************************
 *
 * Distributed under GPL v2.
 *
 * $Id: snmp_common.c,v 1.3 2002/01/27 15:31:50 henning Exp $
 *
 */

#include <limits.h>
#include "snmp_common.h"

char *snmp_version ="$Revision: 1.3 $ ($Date: 2002/01/27 15:31:50 $)";

char *version;         // Version of the program (set in main)
char *bn;              // basename of the program (set in main)

int timeout = 0;       // Timeout for Queries (0 == default)
int verbose = 0;       // Verbosity (0 == not verbose, for production)
int listing = 0;       // Listing mode (0 == default)

char *community = NULL;  // Community, must be set
char *hostname  = NULL;  // Hostname, must be set

/**
 * Print out the command line options
 */

void print_help()
{
  printf("Usage for %s\n", bn);
  printf("-h, --help:                  Display this text\n");
  printf("-V, --version:               Display program version\n");
  printf("-t <secs>, --timeout <secs>: Set program timeout\n");
  printf("-C, --community <name>:      Set SNMP community (required)\n");
  printf("-H, --hostname <name>:       Set Hostname to scan (required)\n");
  printf("-v, --verbose:               Increase verbosity (for testing)\n");
  printf("\nFor check_snmp_disk also applies:\n");
  printf("-l, --list                   List checked partitions\n");
  printf("-p, --path                   Show mount points instead of of devices in list mode\n");
}

/**
 * Print out the program version.
 */

void print_version()
{
  printf("%s: %s\nsnmp_common: %s\n", bn, version, snmp_version);
}

/**
 * Returns 1 if the supplied number is an integer, 0 if not
 */

int is_integer(char *number)
{
  long int n;

  if(strspn(number,"-0123456789 ") != strlen(number))
    return(0);

  n=strtol(number,NULL,10);
  if((errno != ERANGE) && (n >= INT_MIN) && (n <= INT_MAX))
    return(1);

  return(0);
}

/**
 * Fetches an SNMP Table from the supplied OID base.
 *
 * callback gets called for each found element.
 * pnt is a generic pointer argument passed to callback
 * nof is the maximum number of elements to look for. If zero,
 * no limit.
 */

int fetch_table(char *base, callback_t callback, void *pnt, int nof)
{
  int res = 0;

  struct snmp_session  session, *ss;
  struct snmp_pdu *pdu, *response;
  struct variable_list *vars;

  oid root[MAX_OID_LEN];
  oid    name[MAX_OID_LEN];
  size_t rootlen = MAX_OID_LEN;
  size_t namelen;

  int running;
  int status;

  snmp_sess_init( &session );

  init_snmp("disk_plugin");

  if(session.version == SNMP_DEFAULT_VERSION)
    session.version = netsnmp_ds_get_int(NETSNMP_DS_LIBRARY_ID, NETSNMP_DS_LIB_SNMPVERSION);
    
  session.peername = hostname;
  session.community = (unsigned char *)community;
  session.community_len = strlen(community);

  if(timeout)
    session.timeout = timeout  * 1000000L;

  if(!snmp_parse_oid(base, root, &rootlen))
  {
    return -1;
  }
  
  SOCK_STARTUP;
  
  ss = snmp_open(&session);

  if(!ss)
  {
    SOCK_CLEANUP;
    return -1;
  }
  
  memmove(name, root, rootlen * sizeof(oid));
  namelen = rootlen;

  running = -1;

  while(running)
  {
    pdu = snmp_pdu_create(SNMP_MSG_GETNEXT);
    snmp_add_null_var(pdu, name, namelen);

    status = snmp_synch_response(ss, pdu, &response);
    if(status != STAT_SUCCESS)
    {
      if(verbose)
        printf("%s: Error #1\n", bn);
      running = 0;
      res = -1;
      continue;
    }
      
    if(response->errstat != SNMP_ERR_NOERROR)
    {
      running = 0;
      if (response->errstat != SNMP_ERR_NOSUCHNAME)
      {
        if(verbose)
          printf("%s: Error #2\n", bn);
        res = -1;
      }
      continue;
    }
      
    for(vars = response->variables; vars; vars = vars->next_variable)
    {
      if((vars->name_length < rootlen) ||
         (memcmp(root, vars->name, rootlen * sizeof(oid)) != 0))
      {
        running = 0;
        continue;
      }

      if(verbose)
        printf("%s: Found another element\n", bn);


      callback(vars, &pnt);

      res++;
      //
      // We have a maximum count. check it
      //
      if(nof && !(--nof))
      {
        running = 0;
        continue;
      }
      
      if ((vars->type != SNMP_ENDOFMIBVIEW) &&
          (vars->type != SNMP_NOSUCHOBJECT) &&
          (vars->type != SNMP_NOSUCHINSTANCE))
      {
        memmove((char *)name, (char *)vars->name,
                vars->name_length * sizeof(oid));
        
        namelen = vars->name_length;
      }
      else
      {
        running = 0;
      }
    } /* for */

    if(response)
      snmp_free_pdu(response);
  }

  snmp_close(ss);
  SOCK_CLEANUP;
  return res;
}

/**
 * Callback that does nothing. Used for counting the number
 * of elements in a table.
 */

void null_callback(struct variable_list *vars, void **pnt)
{
  if(verbose)
    printf("%s: Null Callback called with %lx\n", bn, vars);
}

/**
 * Stores the integer value of the supplied variable element
 * into pnt. Advances pnt
 */

void integer_callback(struct variable_list *vars, void **pnt)
{
  long *data = (long *)(*pnt);

  if(verbose)
    printf("%s: Found Integer Value %ld into %lx\n", bn, *vars->val.integer, data);

  *data = *vars->val.integer;

  *pnt = ++data;
}

/**
 * Stores the string value of the supplied variable element
 * into pnt. Advances pnt
 */


void string_callback(struct variable_list *vars, void **pnt)
{
  char **data = (char **)(*pnt);
  
  if(verbose)
    printf("%s: Putting String into %lx\n", bn, data);

  if((vars->val.string) && (*data = calloc(sizeof(char), vars->val_len+1)))
  {
    memcpy(*data, vars->val.string, vars->val_len);
  }
  else
  {
    *data = "";
  }

  *pnt = ++data;
}



snmp_common.h
/*************************************************************************
 *
 * plugins for Nagios
 *
 * (C) 2002-2008
 *   Henning P. Schmiedehausen
 *   INTERMETA - Gesellschaft fuer Mehrwertdienste mbH
 *   Hutweide 15
 *   D-91054 Buckenhof
 *
 *************************************************************************
 *
 * Common code
 *
 *************************************************************************
 *
 * Distributed under GPL v2.
 *
 * $Id: snmp_common.h,v 1.2 2002/01/27 15:27:19 henning Exp $
 *
 */

#include "config.h"

#include <sys/types.h>
#include <stdio.h>
#include <libgen.h>

#if HAVE_STDLIB_H
#include <stdlib.h>
#endif

#if HAVE_UNISTD_H
#include <unistd.h>
#endif

#if HAVE_ERRNO_H
#include <errno.h>
#endif

#if HAVE_STRING_H
#include <string.h>
#else
#include <strings.h>
#endif

#if HAVE_GETOPT_H
#define _GNU_SOURCE
#include <getopt.h>
#endif

#ifdef HAVE_NET_SNMP
#include <net-snmp/net-snmp-config.h>
#include <net-snmp/net-snmp-includes.h>
#endif

/*
 * Callback pointer type
 */
typedef void (*callback_t)(struct variable_list *, void **);

/*
 * Routines from snmp_common.c
 */

void print_help(void);
void print_version(void);
int is_integer(char *);
int fetch_table(char *, callback_t, void *, int);

void null_callback   (struct variable_list *, void ** );
void integer_callback(struct variable_list *, void ** );
void string_callback (struct variable_list *, void ** );

/*
 * getopt
 */

extern char *optarg;
extern int opterr;
extern int optopt;
extern int optind;

/*
 * snmp_common Variables
 *
 */

extern char *bn;         // Basename of the program
extern char *version;    // Version of the Program

extern int timeout;      // Timeout (set with -t)
extern int verbose;      // Verbosity (set with -v)
extern int listing;      // List mode (set with -l)
extern char *community;  // SNMP Community (set with -C)
extern char *hostname;   // Hostname (set with -H)

/*
 * From Nagios Include
 *
 */

#define STATE_OK                        0
#define STATE_WARNING                   1
#define STATE_CRITICAL                  2
#define STATE_UNKNOWN                   3       /* changed from -1 on 02/24/2001 */



check_snmp_disk.c
/*************************************************************************
 *
 * plugins for Nagios
 *
 * (C) 2002-2008
 *   Henning P. Schmiedehausen
 *   INTERMETA - Gesellschaft fuer Mehrwertdienste mbH
 *   Hutweide 15
 *   D-91054 Buckenhof
 * (C) 2010-2013
 *   Elan Ruusamäe, <glen@delfi.ee>
 *
 *************************************************************************
 *
 * Checks the disks for a given host via snmp and the
 * ucd snmp interface.
 *
 *************************************************************************
 *
 * Distributed under GPL v2.
 *
 * $Id: disk_plugin.c,v 1.6 2002/01/27 22:10:24 henning Exp $
 *
 */

#include "snmp_common.h"

#define RCS_VERSION "$Revision: 1.6 $ ($Date: 2002/01/27 22:10:24 $)"

#define DISK_INDEX_MIB       ".1.3.6.1.4.1.2021.9.1.1"
#define DISK_PATH_MIB        ".1.3.6.1.4.1.2021.9.1.2"
#define DISK_DEVICE_MIB      ".1.3.6.1.4.1.2021.9.1.3"
#define DISK_MINIMUM_MIB     ".1.3.6.1.4.1.2021.9.1.4"
#define DISK_MINPERCENT_MIB  ".1.3.6.1.4.1.2021.9.1.5"
#define DISK_TOTAL_MIB       ".1.3.6.1.4.1.2021.9.1.6"
#define DISK_AVAIL_MIB       ".1.3.6.1.4.1.2021.9.1.7"
#define DISK_USED_MIB        ".1.3.6.1.4.1.2021.9.1.8"
#define DISK_PERCENT_MIB     ".1.3.6.1.4.1.2021.9.1.9"
#define DISK_PERCENTNODE_MIB ".1.3.6.1.4.1.2021.9.1.10"
#define DISK_ERRORFLAG_MIB   ".1.3.6.1.4.1.2021.9.1.100"
#define DISK_ERRORMSG_MIB    ".1.3.6.1.4.1.2021.9.1.101"

int report_disk(void);

int path = 0;       // path mode (0 == default device mode)

int main (int argc, char *argv[])
{
  static struct option long_options[] = {
    { "help",      no_argument,       0, 'h' },
    { "version",   no_argument,       0, 'V' },
    { "timeout",   required_argument, 0, 't' },
    { "community", required_argument, 0, 'C' },
    { "hostname",  required_argument, 0, 'H' },
    { "verbose",   no_argument,       0, 'v' },
    { "list",      no_argument,       0, 'l' },
    { "path",      no_argument,       0, 'p' },
    { 0, 0, 0, 0 },
  };
  int option_index = 0;
  int c;

  int ret = STATE_UNKNOWN;

  bn = strdup(basename(argv[0]));
  version = VERSION;

#define OPTS "?hVvlpt:c:w:C:H:"
  
  while(1)
  {
    c = getopt_long(argc, argv, OPTS, long_options, &option_index);

    if(c == -1 || c == EOF)
      break;

    switch(c)
    {
      case '?':
      case 'h':
        print_help();
        exit(STATE_UNKNOWN);

      case 'V':
        print_version();
        exit(STATE_UNKNOWN);


      case 't':
        if(!is_integer(optarg))
        {
          printf("%s: Timeout interval (%s)must be integer!\n",
                 bn,
                 optarg);
          exit(STATE_UNKNOWN);
        }
        
        timeout = atoi(optarg);
        if(verbose)
          printf("%s: Timeout set to %d\n", bn, timeout);
        break;

      case 'C':
        community = strdup(optarg);

        if(verbose)
          printf("%s: Community set to %s\n", bn, community);
        
        break;

      case 'H':
        hostname = strdup(optarg);

        if(verbose)
          printf("%s: Hostname set to %s\n", bn, hostname);

        break;

      case 'v':
        verbose = 1;
        printf("%s: Verbose mode activated\n", bn);
        break;

      case 'l':
        listing = 1;
        if(verbose)
            printf("%s: List mode activated\n", bn);
        break;

      case 'p':
        path = 1;
        if(verbose)
            printf("%s: Listing mount point activated\n", bn);
        break;
    }
  }

  if(!hostname || !community)
  {
    printf("%s: Hostname or Community missing!\n", bn);
    print_help();
    exit(STATE_UNKNOWN);
  }

  ret = report_disk();

  if(verbose)
    printf("%s: Returning %d\n", bn, ret);

  exit(ret);
}

int report_disk()
{
  int cnt;

  long *errors;
  void *pnt;
  int i;
  int gotErrors = 0;
  char **errormsg = NULL;
  char **diskname = NULL;

  
  if((cnt = fetch_table(DISK_INDEX_MIB, null_callback, NULL, 0)) < 0)
  {
    printf("%s: Could not fetch list of disks\n", bn);
    return STATE_UNKNOWN;
  }

  if(!cnt)  // No disks configured
  {
    printf("%s: No disks found.\n", bn);
    return STATE_WARNING;
  }
  

  if(!(errors  = calloc(sizeof(long), cnt)))
  {
    printf("%s: Could not allocate memory for information\n", bn);
    return STATE_CRITICAL;
  }

  pnt = errors;
  if(fetch_table(DISK_ERRORFLAG_MIB, integer_callback, pnt, cnt) < 0)
  {
    printf("%s: Could not fetch error list!\n", bn);
    return STATE_UNKNOWN;
  }

  for(i=0; i<cnt; i++)
  {
    if(verbose)
      printf("%s: Got Flag %ld for %d\n", bn, errors[i], i);

    if(errors[i])
    {
      gotErrors = 1;
      break;
    }
  }

  errormsg = calloc(sizeof(char **), cnt);
  diskname = calloc(sizeof(char **), cnt);
  if(!errormsg || !diskname)
  {
    printf("%s: Could not allocate memory for error information\n", bn);
    return STATE_CRITICAL;
  }

  pnt = errormsg;

  if(fetch_table(DISK_ERRORMSG_MIB, string_callback, pnt, cnt) < 0)
  {
    printf("%s: Could not fetch error messages\n", bn);
    return STATE_CRITICAL;
  }

  pnt = diskname;

  if(fetch_table(path ? DISK_PATH_MIB : DISK_DEVICE_MIB, string_callback, pnt, cnt) < 0)
  {
    printf("%s: Could not fetch device list\n", bn);
    return STATE_CRITICAL;
  }

  if(gotErrors == 1) {
    for(i=0; i < cnt; i++)
    {
      if(errors[i])
        printf("%s (%s); ", errormsg[i], diskname[i]);
    }
    printf("\n");
  } else {
    if(listing)
    {
      printf( "Checked %d disks (%s", cnt, diskname[0]);
      for(i=1; i < cnt; i++)
      {
          printf( ", %s", diskname[i] );
      }
      printf(").\n" );
    }
    else {
      printf("Checked %d disks.\n", cnt);
    }
    return STATE_OK;
  }
  
  return STATE_CRITICAL;
}

//Setup VIM: ex: et ts=2 sw=2 enc=utf-8 :


check_snmp_proc.c
/*************************************************************************
 *
 * plugins for Nagios
 *
 * (C) 2002-2008
 *   Henning P. Schmiedehausen
 *   INTERMETA - Gesellschaft fuer Mehrwertdienste mbH
 *   Hutweide 15
 *   D-91054 Buckenhof
 * (C) 2007
 *   Elan Ruusamäe, <glen@delfi.ee>
 *
 *************************************************************************
 *
 * Checks the running processes for a given host via snmp and the
 * ucd snmp interface.
 *
 *************************************************************************
 *
 * Distributed under GPL v2.
 *
 * $Id: proc_plugin.c,v 1.6 2002/01/27 22:10:24 henning Exp $
 *
 */

#include "snmp_common.h"

#define RCS_VERSION "$Revision: 1.6 $ ($Date: 2002/01/27 22:10:24 $)"

#define PROC_INDEX_MIB       ".1.3.6.1.4.1.2021.2.1.1"
#define PROC_NAME_MIB        ".1.3.6.1.4.1.2021.2.1.2"
#define PROC_MIN_MIB         ".1.3.6.1.4.1.2021.2.1.3"
#define PROC_MAX_MIB         ".1.3.6.1.4.1.2021.2.1.4"
#define PROC_RUNNING_MIB     ".1.3.6.1.4.1.2021.2.1.5"
#define PROC_ERRORFLAG_MIB   ".1.3.6.1.4.1.2021.2.1.100"
#define PROC_ERRORMSG_MIB    ".1.3.6.1.4.1.2021.2.1.101"

int report_proc(void);

int main (int argc, char *argv[])
{
  static struct option long_options[] = {
    { "help",      no_argument,       0, 'h' },
    { "version",   no_argument,       0, 'V' },
    { "timeout",   required_argument, 0, 't' },
    { "community", required_argument, 0, 'C' },
    { "hostname",  required_argument, 0, 'H' },
    { "verbose",   no_argument,       0, 'v' },
    { "list",      no_argument,       0, 'l' },
    { 0, 0, 0, 0 },
  };
  int option_index = 0;
  int c;

  int ret = STATE_UNKNOWN;

  bn = strdup(basename(argv[0]));
  version = VERSION;

#define OPTS "?hVvlt:c:w:C:H:"
  
  while(1)
  {
    c = getopt_long(argc, argv, OPTS, long_options, &option_index);

    if(c == -1 || c == EOF)
      break;

    switch(c)
    {
      case '?':
      case 'h':
        print_help();
        exit(STATE_UNKNOWN);

      case 'V':
        print_version();
        exit(STATE_UNKNOWN);


      case 't':
        if(!is_integer(optarg))
        {
          printf("%s: Timeout interval (%s)must be integer!\n",
                 bn,
                 optarg);
          exit(STATE_UNKNOWN);
        }
        
        timeout = atoi(optarg);
        if(verbose)
          printf("%s: Timeout set to %d\n", bn, timeout);
        break;

      case 'C':
        community = strdup(optarg);

        if(verbose)
          printf("%s: Community set to %s\n", bn, community);
        
        break;

      case 'H':
        hostname = strdup(optarg);

        if(verbose)
          printf("%s: Hostname set to %s\n", bn, hostname);

        break;

      case 'v':
        verbose = 1;
        printf("%s: Verbose mode activated\n", bn);
        break;

      case 'l':
        listing = 1;
		if(verbose)
			printf("%s: List mode activated\n", bn);
        break;
    }
  }

  if(!hostname || !community)
  {
    printf("%s: Hostname or Community missing!\n", bn);
    print_help();
    exit(STATE_UNKNOWN);
  }

  ret = report_proc();

  if(verbose)
    printf("%s: Returning %d\n", bn, ret);

  exit(ret);
}

void format_report(int cnt, long *errors, char **errormsg)
{
  int i;
  size_t nmany = 0, nfew = 0, nunkn = 0;
  char **errmany = NULL, **errfew = NULL, **errunkn = NULL;
  char **pmany = NULL, **pfew = NULL, **punkn = NULL;

  errmany = calloc(sizeof(char **), cnt);
  if (!errmany) {
    printf("%s: Could not allocate memory for error information\n", bn);
    return;
  }
  errfew = calloc(sizeof(char **), cnt);
  if (!errfew) {
    printf("%s: Could not allocate memory for error information\n", bn);
    return;
  }
  errunkn = calloc(sizeof(char **), cnt);
  if (!errunkn) {
    printf("%s: Could not allocate memory for error information\n", bn);
    return;
  }

#define _msg_few "Too few "
#define _msg_many "Too many "
  pmany = errmany; pfew = errfew; punkn = errunkn;
  for (i = 0; i < cnt; i++) {
    if (errors[i]) {
      size_t len = strlen(errormsg[i]);
      if (strncmp(errormsg[i], _msg_many, sizeof(_msg_many) - 1) == 0) {
        *(pmany++) = errormsg[i] + sizeof(_msg_many) - 1;
        nmany++;

      } else if (strncmp(errormsg[i], _msg_few, sizeof(_msg_few) - 1) == 0) {
        *(pfew++) = errormsg[i] + sizeof(_msg_few) - 1;
        nfew++;

      } else {
        *(punkn++) = errormsg[i];
        nunkn++;
      }
    }
  }

  if (verbose) {
    printf("%s: Got %ld few, %ld many, %ld unknown messages\n", bn, nfew, nmany, nunkn);
  }

#define _msg_running " running"
  if (nmany) {
    char *p;
    printf("%s", _msg_many);
    for (i = 0; i < nmany- 1; i++) {
      p = strstr(errmany[i], _msg_running);
      printf("%.*s%s, ", p - errmany[i], errmany[i], p + sizeof(_msg_running) - 1);
    }
    p = strstr(errmany[i], _msg_running);
    printf("%.*s%s", p - errmany[i], errmany[i], p + sizeof(_msg_running) - 1);
    printf("%s", _msg_running);
  }
  if (nfew) {
    char *p;
    if (nmany) {
      printf(". ");
    }
    printf("%s", _msg_few);
    for (i = 0; i < nfew - 1; i++) {
      p = strstr(errfew[i], _msg_running);
      printf("%.*s%s, ", p - errfew[i], errfew[i], p + sizeof(_msg_running) - 1);
    }
    p = strstr(errfew[i], _msg_running);
    printf("%.*s%s", p - errfew[i], errfew[i], p + sizeof(_msg_running) - 1);
    printf("%s", _msg_running);
  }
  if (nunkn) {
    if (nfew) {
      printf(". ");
    }
    for (i = 0; i < nunkn - 1; i++) {
      printf("%s", errunkn[i]);
    }
    printf("%s", errunkn[i]);
  }
  printf("\n");
}

int report_proc()
{
  int cnt;

  long *errors;
  void *pnt;
  int i;
  int gotErrors = 0;
  char **errormsg = NULL;
  char **procname = NULL;
  
  if((cnt = fetch_table(PROC_INDEX_MIB, null_callback, NULL, 0)) < 0)
  {
    printf("%s: Could not fetch list of processes\n", bn);
    return STATE_UNKNOWN;
  }

  if(!cnt)  // No procs configured
  {
    printf("%s: No processes found.\n", bn);
    return STATE_OK;
  }

  if(!(errors  = calloc(sizeof(long), cnt)))
  {
    printf("%s: Could not allocate memory for information\n", bn);
    return STATE_CRITICAL;
  }

  procname = calloc(sizeof(char **), cnt);

  pnt = errors;
  if(fetch_table(PROC_ERRORFLAG_MIB, integer_callback, pnt, cnt) < 0)
  {
    printf("%s: Could not fetch error list!\n", bn);
    return STATE_UNKNOWN;
  }

  for(i=0; i<cnt; i++)
  {
    if(verbose)
      printf("%s: Got Flag %ld for %d\n", bn, errors[i], i);

    if(errors[i])
    {
      gotErrors = 1;
      break;
    }
  }

	pnt = procname;
	if(fetch_table(PROC_NAME_MIB, string_callback, pnt, cnt) < 0)
	{
		printf("%s: Could not fetch process list\n", bn);
		return STATE_CRITICAL;
	}

  if(gotErrors == 0)
  {
	if(listing)
	{
		printf("Checked %d process groups (%s", cnt, procname[0]);
		for(i=1; i < cnt; i++)
		{
			printf( ", %s", procname[i] );
		}
		printf( ").\n");
	}
	else
    	printf("Checked %d process groups.\n", cnt);
    return STATE_OK;
  }

  errormsg = calloc(sizeof(char **), cnt);
  if(!errormsg)
  {
    printf("%s: Could not allocate memory for error information\n", bn);
    return STATE_CRITICAL;
  }
  
  pnt = errormsg;
  
  if(fetch_table(PROC_ERRORMSG_MIB, string_callback, pnt, cnt) < 0)
  {
    printf("%s: Could not fetch error messages\n", bn);
    return STATE_CRITICAL;
  }

  format_report(cnt, errors, errormsg);

  return STATE_CRITICAL;
}
